editor/components/document-outline/index.js

```diff
@@ -57,10 +57,24 @@ const getHeadingLevel = heading => {
 	}
 };
 
+const computeOutlineHeadings = ( blocks = [], path = [] ) => {
+	return flatMap( blocks, ( block = {} ) => {
+		if ( block.name === 'core/heading' ) {
+			return {
+				...block,
+				path,
+				level: getHeadingLevel( block ),
+				isEmpty: isEmptyHeading( block ),
+			};
+		}
+		return computeOutlineHeadings( block.innerBlocks, [ ...path, block ] );
+	} );
+};
+
 const isEmptyHeading = heading => ! heading.attributes.content || heading.attributes.content.length === 0;
 
 export const DocumentOutline = ( { blocks = [], title, onSelect } ) => {
-	const headings = filter( blocks, ( block ) => block.name === 'core/heading' );
+	const headings = computeOutlineHeadings( blocks );
```

Comment:
As discussed on Slack today. Can we convert this functionality (`computeOutlineHeadings `) into a general purpose selector `getBlocksOfType`?

Reply:
Unfortunately, at the time of our discussion, I did not remember that here we don't want just the heading blocks we also want the path to it in order to display it to the user.
The path is the array of ancestors of a block until a heading. so it can be ["Core/collumn"] or [] or ["cover/image", "Core/collumn"].
I think this is the only place where we need to retrieve the blocks of a given type sorted by document position with its ancestors so this function is probably not reusable elsewhere.